### PR TITLE
Add code to create the different ILLink packages

### PR DESCRIPTION
**Pull request by tlakollo**
With these changes, we get rid of the Packable logic from dotnet/linker and use the .pkgproj logic that is used by NativeAOT
* Add tools subset and set it to execute tools.illink by default
* Add the tools subset to the DefaultSubsets in runtime to build illink by default
* Add the illink-packages.proj as a project to build when executing build.cmd packs
* Add logic to generate the different packages using the illink-packages.proj and different .pkgproj
* Remove packable logic
* Re-add TargetFrameworks to ILLink.Tasks and set TargetFramework to empty otherwise it will complain about setting the TargetFramework twice

### CODE REVIEW

File: src/installer/pkg/projects/Microsoft.NET.ILLink/Microsoft.NET.ILLink.pkgproj [hunk 1]
```diff
@@ -0,0 +1,21 @@
+<Project Sdk="Microsoft.NET.Sdk">
```
Comment (ViktorHofer) [inline src/installer/pkg/projects/Microsoft.NET.ILLink/Microsoft.NET.ILLink.pkgproj:1]: We must not introduce new pkgproj projects: https://github.com/dotnet/runtime/issues/75966. Please use a NuGet Pack task based project instead.

**Review by ViktorHofer — CHANGES_REQUESTED** [review]
As mentioned, we don't want to use a pkgproj here and instead want the ILLink.Tasks.csproj project to use the NuGet Pack task functionality. Pkgproj (which is a non shipping, internal packaging infrastructure) is something that we want to get rid off and that we don't support anymore.

**Comment by agocke** [comment]
> As mentioned, we don't want to use a pkgproj here and instead want the ILLink.Tasks.csproj project to use the NuGet Pack task functionality. Pkgproj (which is a non shipping, internal packaging infrastructure) is something that we want to get rid off and that we don't support anymore.

Oh, that's news to me. I've tried to make use of the built-in NuGet functionality recently for some complex packages and ended up going with https://github.com/devlooped/nugetizer because it's too hard to put multiple references in one package. That seems like a common case for the runtime repo. How will we work around that?

**Comment by ViktorHofer** [comment]
> Oh, that's news to me. I've tried to make use of the built-in NuGet functionality recently for some complex packages and ended up going with https://github.com/devlooped/nugetizer because it's too hard to put multiple references in one package. That seems like a common case for the runtime repo. How will we work around that?

We brought this to the M2s attention during ~October/November. You likely were out during that time.
Can you please describe the requirements for this package in more detail? What is the exact package layout that you want to have?

Many of our packages produced under src/libraries have multiple assemblies in them (ie source generators or transport packages) and leverage extensions that sit on top of the NuGet Pack task that we bundled into a common file: [packaging.targets](https://github.com/dotnet/runtime/blob/b1a2080b9ce5833802fe2d632f79de6402097f14/eng/packaging.targets#L231).

**Comment by ViktorHofer** [comment]
> My ideal state is that NuGet would add direct support for package projects of some kind

That is tracked via https://github.com/NuGet/Home/issues/3891

> but in the intervening time I actually think using something like Nugetizer would be a pretty good alternative

Is Nugetizer source-buildable? I don't think that any of our components in the stack use it today.

> The file you pointed to seems neat, it looks like it makes some stuff easier. But it doesn't change the situation that having a separate project for the packages is often the best way of doing things. The advantages of package projects are:

While some of your points are nice to have, the disadvantages that drove us away from pkgproj by far outweigh its benefits:
- We don't want to develop, maintain and service our own packaging infrastructure when there's already a team (NuGet) that offers a decent solution to bundle project specific or arbitrary assets into a package.
- We want to leverage prebuilts via `PackageReference` items and pkgprojs aren't designed to be restore-able.
- We want a strong dependency between the build and the package authoring of a component so that a developer's innerloop is as efficient as possible: `dotnet pack`
- We want to follow the workflow that's already established in the .NET ecosystem: `dotnet pack`, `GeneratePackageOnBuild`, `PackageId`, `PackageDescription`, `...`

> In both of these cases external contributors have to learn about our custom extensions to understand our packaging.

The extension points that our packaging.targets file uses are the officially documented NuGet Pack task hook points that customers use in their real world applications: https://learn.microsoft.com/en-us/nuget/reference/msbuild-targets#advanced-extension-points-to-create-customized-package. I don't think that we re-invent the wheel with our current packaging infrastructure, especially in contrast to what we had before with pkgproj which was a direct copy of the NuGet packaging related C# code.

I'm willing to help make these projects follow our best practices (use NuGet Pack task, avoid pkgproj). @tlakollo already reached out to me offline yesterday and we will sync up today. Please let me know if you want this to go a different direction. I'm more or less fine with any solution that doesn't depend on the pkgproj infrastructure as we plan to remove that sooner than later.

cc @ericstj

**Comment by agocke** [comment]
> The extension points that our packaging.targets file uses are the officially documented NuGet Pack task hook points that customers use in their real world applications: https://learn.microsoft.com/en-us/nuget/reference/msbuild-targets#advanced-extension-points-to-create-customized-package. I don't think that we re-invent the wheel with our current packaging infrastructure, especially in contrast to what we had before with pkgproj which was a direct copy of the NuGet packaging related C# code.

You're right, this is a big improvement over the level of hacks we were doing for pkgproj. Still, there are customizations that authors need to learn.

> I'm willing to help make these projects follow our best practices (use NuGet Pack task, avoid pkgproj). @tlakollo already reached out to me offline yesterday and we will sync up today.

Thanks for helping out.

My big takeaway here: we need to have documentation for each of these extensions so that people can understand what our packaging stuff does and how to write new packages. Could you write up some docs, or point me to the existing ones?

**Comment by ViktorHofer** [comment]
> My big takeaway here: we need to have documentation for each of these extensions so that people can understand what our packaging stuff does and how to write new packages. Could you write up some docs, or point me to the existing ones?

We have https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/libraries-packaging.md which covers most of the available extensions.

**Pull request closed without merging**